accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,354 @@
+  
+and has limitations and issues:
+- Cannot be used with byref-like types like `Span<T>` either has the target or as an argument. This is because by-ref like types cannot be boxed. **This is the key limitation expressed in this document.**
+- `ref` and `out` parameters are retrieved after `Invoke()` through the `parameters` argument. This is a manual mechanism performed by the user and means there is no argument or return value "aliasing" to the original variable.
+- Boxing of value types makes it not possible to invoke a mutable method and have the target `obj` parameter updated.
+- [`System.Reflection.Pointer.Box()`](https://learn.microsoft.com/dotnet/api/system.reflection.pointer.box?view=net-7.0) and `UnBox()` must be used to manually box and unbox a pointer (`*`) type.
+- When an exception occurs in the target method, the exception is wrapped with a `TargetInvocationException` and re-thrown. This approach is not desired in most cases and somewhat recently, `BindingFlags.DoNotWrapExceptions` was added to change this behavior as an opt-in. Not having a `try\catch` would help a bit with performance as well.
+
+Due to the performance issues and usability issues, workarounds and alternatives are used including:
+- [MethodInfo.CreateDelegate()](https://learn.microsoft.com/dotnet/api/system.reflection.methodinfo.createdelegate) which supports a direct method invocation and thus is strongly-typed and not appropriate for loosely-typed invoke scenarios.
+- [Dynamic methods](https://learn.microsoft.com/dotnet/framework/reflection-and-codedom/how-to-define-and-execute-dynamic-methods) which are IL-emit based and include a non-trivial implementation.
+- Compiled [expression trees](https://learn.microsoft.com/dotnet/csharp/programming-guide/concepts/expression-trees/) which use dynamic method if IL Emit is available but also conveniently fallback to standard reflection when IL Emit is not available.
+
+# 8.0 Goals
+The 7.0 release had a 3-4x perf improvement for `Invoke()` by using IL Emit when available and falling back to standard reflection when not available, keeping the same object-based APIs. Although this improvement is significant, it doesn't replace the need to use the IL-Emit based alternatives for the highly-performance-sensitive scenarios such as the `System.Text.Json` serializer. New APIs are required.
+
+For 8.0, there are two primary goals:
+1) Support byref-like types both for invoking and passing as arguments. An unsafe approach may need to be used for V8. This unblocks scenarios.
+2) Support "fast invoke" so that dynamic methods no longer have to be used for performance. Today both `System.Text.Json` and [dependency injection](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection) use IL emit for performance. As a litmus test, these areas will be changed to use the new APIs proposed here.
```

Comment:
Replacing dynamic methods in `Microsoft.Extensions.DependencyInjection` is more important than replacing dynamic methods in `System.Text.Json`.

Our .NET 8 plan is to keep `Microsoft.Extensions.DependencyInjection` dependency on reflection for trimming and native AOT compatibility. It means that we really need a fast reflection invoke that does not depend on dynamic code generation.

The plan for System.Text.Json is to use source generators for trimming and native AOT compatibility. It means that we do need to get rid of the dynamic code generation in (non-source generated) System.Text.Json to achieve our .NET 8 goals.

cc @eerhardt @davidfowl

Reply:
Right. Today, it's hard to remove reflection from dependency injection because the resolution of instances relies on the knowledge that only exists at runtime (the list of services registered in the container). We could do partial codegen for each module but that could end up being slower (it's similar to partial crossgen).

This is the implementation that gets used at runtime when resolving instances in AOT mode using DI:

https://github.com/dotnet/runtime/blob/5b8ebeabb32f7f4118d0cc8b8db28705b62469ee/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs

Reply:
> It means that we really need a fast reflection invoke that does not depend on dynamic code generation and that supports everything that Microsoft.Extensions.DependencyInjection needs.

To clarify: for "fast invoke", the new APIs still require "dynamic code generation" through IL Emit but of course use the "slow invoke" interpreted fallback approach when not available. So DI would use the new reflection APIs instead instead of expressions. An unstated goal is to support https://github.com/dotnet/runtime/issues/66153 which doesn't want to use expressions in Blazor client due to having to bring along the large System.Linq.Expressions.dll.

Reply:
>  for "fast invoke", the new APIs still require "dynamic code generation" through IL Emit but of course use the "slow invoke" interpreted fallback approach when not available. 

For native AOT config, the invocation should be done via a stub generated at build time. This stub is similar to the stub generated via IL Emit and it has similar performance. It is how reflection invoke works in native AOT today. The slow interpreted fallback does not exist in native AOT.

Reply:
it seems like we could delete a large amount of code that uses expression trees in some cases if we could have the CoreCLR version of these APIs use reflection emit under the covers. Then we could just use the built in methods and have them be "best in class" on each of the available platforms.

Reply:
> I'll assume for now "InvokeDirect()` is as fast as possible with perhaps only argument number validated

The set\get parameters have to store the additional type information so that it can maybe validated later. Is that correct? `InvokeDirect` maybe as fast as possible, but the rest is not.

Reply:
> The set\get parameters have to store the additional type information so that it can maybe validated later. Is that correct?

Not when using the `object` set overloads. The invoker maintains a tuple **(byreference,** **object,** **type)** for each arg and depending on the set method used and whether the target method has byref parameters. For the simple `object`-only case used by DI, for example, we don't need the `byreference` or the `type` for Core at least.

For perf, Core will have both the byref- and object-based emit. NativeAOT may elect to force everything to byrefs to have just one stub.

Reply:
For DI I'll provide more detail, but given this perf test:
- DI factory logic which is `object`-based with no byref args
- The perf measures calling the factory method. This was earlier obtained from `ActivatorUtilities.CreateFactory(Type, Type[])` but that is not included in the benchmark.
- That factory method was changed to use the new `MethodInvoker`.
- The factory method has 3 parameters, 2 of which are "matched" and 1 is "unmatched" so there's some logic involved in calling the IServiceProvider for the unmatched case.

The breakdown of time spent:
- ~12% for calling the invoke delegate which call the ctor of the target type passing the 3 args.
- ~5% for MethodInvoker.Invoke() overhead
- ~80% for setting the args and preparing to call MethodInvoker.Invoke(). This can vary depending on fixed vs. variable MethodInvoker APIs, but not much. This also includes the DI logic for default values and mapping.

That 80% dwarfs the actual invoke of 12%.

The test ran in ~55ns compared to ~66ns for today's reflection (which uses emit since 7.0). This ~17% gain also includes some other misc optimizations. In summary, this is not getting us much except there is no more alloc for the `object[]` for the args.

Compare this to LINQ with compiled expressions which is ~2.5x faster (~22ns).

Based on this I have a new prototype:
- Added additional Invoke() APIs for fixed-length cases (say up to 8). Added a static MethodInvoker.Invoke(...) or MethodInvoker.GetInvokeDelegate(...) that requires no allocs or setting of values. It will only take `object` args and be focused around existing `MethodBase.Invoke` cases which can only take `object`.
- Perform a manual "AOT" generation of the DI factory lambda code depending how the matching of args occurs. This is based on whether default values are required and the mapping logic. It avoids capturing unnecessary state by the factory's lambda and avoids various logic to check for default values etc.

This new prototype looks promising but will never be as fast as LINQ expressions since that can hard-code the default values and mapping information in the expression. For the happy path, I think I can get it to within 30-50% (so will still be 1.3-1.5x slower, but not 2.5x).